File: src/main/java/lambdaspecial/java/QuickSortLambda.java
package lambdaspecial.java;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class QuickSortLambda<T extends Comparable<T>> implements Sort<T> {

    @Override
    public List<T> sort(List<T> array) {
        if (array == null || array.size() < 2) {
            return array;
        }

        T pivot = array.get(array.size() / 2);

        Map<Integer,List<T>> grouped = array.stream()
                .collect(
                        Collectors.groupingBy(e -> Integer.compare(e.compareTo(pivot), 0),
                        Collectors.mapping(e -> e, Collectors.toList())));

        var less = grouped.get(-1);
        var equal = grouped.get(0);
        var greater = grouped.get(1);
        return Stream.of(sort(less), equal, sort(greater))
                    .filter(Objects::nonNull)
                    .flatMap(List::stream)
                    .toList();
    }
}

File: src/main/kotlin/lambdaspecial/kotlin/QuickSortLambda.kt
package lambdaspecial.kotlin

class QuickSortLambda<T : Comparable<T>> : Sort<T> {

    override fun sort(array: List<T>): List<T> {

        if (array.size <= 1) return array

        val pivot = array[array.size / 2]

        val grouped = array.groupBy {
            when {
                it < pivot -> -1
                it == pivot -> 0
                else -> 1
            }
        }

        val less = grouped[-1]?: emptyList()
        val equal = grouped[0]?: emptyList()
        val greater = grouped[1]?: emptyList()

        return sequenceOf(sort(less), equal, sort(greater))
                .flatMap { it.asSequence() }
                .toList()
    }
}
